activemodel/lib/active_model/type/collection.rb

```diff
@@ -0,0 +1,76 @@
+# frozen_string_literal: true
+
+module ActiveModel
+  module Type
+    # Attribute type for a collection of values. It is registered under
+    # the +:collection+ key. +:element_type+ option is used to specify elements type
+    #
+    #   class User
+    #     include ActiveModel::Attributes
+    #
+    #     attribute :lucky_numbers, :collection, element_type: :integer
+    #   end
+    #
+    #   user = User.new(lucky_numbers: [1, 2, 3])
+    #   user.lucky_numbers # => [1, 2, 3]
+    #
+    # Value is wrapped into an Array if not an Array already
+    #
+    #   User.new(lucky_numbers: 1).lucky_numbers # => [1]
+    #
+    # Collection elements are coerced by their +:element_type+ type
+    #
+    #   User.new(lucky_numbers: ["1"]).lucky_numbers # => [1]
+    #
+    class Collection < Value
+      def initialize(**args)
+        @element_type = args.delete(:element_type)
+        @type_object = Type.lookup(element_type, **args)
+        @serializer = args.delete(:serializer) || ActiveSupport::JSON
+        super()
+      end
+
+      def type
+        :collection
+      end
+
+      def cast(value)
+        return [] if value.nil?
+        Array(value).map { |el| @type_object.cast(el) }
+      end
+
+      def serializable?(value)
+        value.all? { |el| @type_object.serializable?(el) }
+      end
+
+      def serialize(value)
+        serializer.encode(value.map { |el| @type_object.serialize(el) })
```

Comment:
While testing how this type would work with `element_type:  :model`, I came across a potential issue regarding the serialization of model collections.

Let's say that we have the following models:
```ruby
class Book
  inlcude ActiveModel::Model
  include ActiveModel::Attributes

  attribute :chapters, :collection, element_type: :model, class_name: "Book::Chapter"
end

class Book::Chapter
  inlcude ActiveModel::Model
  include ActiveModel::Attributes

  attribute :name, :string
end
```
When calling `book.attributes_for_database["chapters"]`, we'll get an output that looks like this:
```ruby
"[\"{\\\"name\\\":\\\"Chapter 1\\\"}\",\"{\\\"name\\\":\\\"Chapter 2\\\"}\",\"{\\\"name\\\":\\\"Chapter 3\\\"}\"]"
```
After changing the code to this:
```suggestion
        attributes = value.map do |el|
          if element_type == :model
            el.attributes
          else
            @type_object.serialize(el)
          end
        end

        serializer.encode(attributes)  
```
We'll get a properly serialized output:
```ruby
"[{\"name\":\"Chapter 1\"},{\"name\":\"Chapter 2\"},{\"name\":\"Chapter 3\"}]"
```

Reply:
Oh, that's an interesting catch 👀 
I think it was making sense from tech design point of view and as long as `deserialize` implementation is 
https://github.com/rails/rails/blob/96602f40eb2f29b6925c2528f369ec2d8f4ec5b0/activemodel/lib/active_model/type/collection.rb#L51
So by meaning we expect `value` to be a serialized array of serialized values, so when we deserialize an array we still get an array of serialized values which can be deserialized by the collection type

However now I feel like it may not make a lot of sense on practice 🤔 `deserialize` would need to be changed with special handling for `:model` though

I need to think a bit more on this, thanks for bringing this up!